gradoop-common/src/main/java/org/gradoop/common/model/impl/properties/PropertyValueComparator.java

```diff
@@ -0,0 +1,277 @@
+      if (compare != 0) {
+        return compare;
+      }
+    }
+    return 0;
+  }
+
+  /**
+   * Compare maps of property values.
+   * Maps are compared according to the following criteria, in order:
+   * <ol>
+   *   <li>The map size.</li>
+   *   <li>Their key sets. (Compared like other property sets.)</li>
+   *   <li>Their value sets. (Ordered by key.)</li>
+   * </ol>
+   *
+   * @param left  The first map.
+   * @param right The second map.
+   * @return The comparison result.
+   */
+  private int compareMap(Map<PropertyValue, PropertyValue> left,
+    Map<PropertyValue, PropertyValue> right) {
+    final int sizeLeft = left.size();
+    final int sizeRight = right.size();
+    if (sizeLeft != sizeRight) {
+      return Integer.compare(sizeLeft, sizeRight);
+    }
+    List<PropertyValue> keysLeft = new ArrayList<>(left.keySet());
+    keysLeft.sort(this);
+    List<PropertyValue> keysRight = new ArrayList<>(right.keySet());
+    keysRight.sort(this);
+    final int keyCompareResult = compareList(keysLeft, keysRight);
+    if (keyCompareResult != 0) {
+      return keyCompareResult;
+    }
+    for (int i = 0; i < sizeLeft; i++) {
+      final PropertyValue leftVal = left.get(keysLeft.get(i));
+      final PropertyValue rightVal = right.get(keysRight.get(i));
+      final int valCompareResult = compare(leftVal, rightVal);
+      if (valCompareResult != 0) {
+        return valCompareResult;
+      }
+    }
+    return 0;
+  }
+
+  /**
+   * Convert a property value of the {@link Kind#COLLECTION} kind to a {@link List}.
+   *
+   * @param value The property value.
+   * @return The list.
+   */
+  private List<PropertyValue> getAsList(PropertyValue value) {
+    if (value.isList()) {
+      return value.getList();
+    } else if (value.isSet()) {
+      List<PropertyValue> list = new ArrayList<>(value.getSet());
+      list.sort(this);
+      return list;
+    } else {
+      throw new IllegalStateException(
+        "The type of the property value was not determined correctly.");
+    }
+  }
+
+  /**
+   * Convert a property value of the {@link Kind#DATE} kind to a {@link LocalDateTime}.
+   *
+   * @param value The property value.
+   * @return The date.
+   */
+  private LocalDateTime getAsLocalDataTime(PropertyValue value) {
+    if (value.isDate()) {
+      return value.getDate().atStartOfDay();
+    } else if (value.isDateTime()) {
+      return value.getDateTime();
+    } else {
+      throw new IllegalStateException(
+        "The type of the property value was not determined correctly.");
+    }
+  }
+
+  /**
+   * Get the kind of a property value.
+   *
+   * @param value The property value.
+   * @return Its kind.
+   * @throws ClassCastException When the type could not be determined. This is necessary to keep
+   * this implementation consistent with {@link Comparable}.
+   */
+  private Kind getKind(PropertyValue value) {
+    if (value == null || value.equals(PropertyValue.NULL_VALUE)) {
```

Comment:
You are right, Comparators can accept `null`, but the `compareTo` method of `Comparable` should throw an `NullPointerException`. I will change it, since this comparator is used in a `compareTo` method.